src/prompt-input/internal.tsx

```diff
@@ -0,0 +1,168 @@
+// SPDX-License-Identifier: Apache-2.0
+import React, { Ref, useCallback, useEffect, useRef } from 'react';
+import { getBaseProps } from '../internal/base-component';
+import InternalButton from '../button/internal';
+import { fireKeyboardEvent, fireNonCancelableEvent } from '../internal/events';
+import { PromptInputProps } from './interfaces';
+import { useFormFieldContext } from '../internal/context/form-field-context';
+import useForwardFocus from '../internal/hooks/forward-focus';
+import clsx from 'clsx';
+import styles from './styles.css.js';
+import { InternalBaseComponentProps } from '../internal/hooks/use-base-component';
+import { convertAutoComplete } from '../input/utils';
+import { useDensityMode } from '@cloudscape-design/component-toolkit/internal';
+
+export interface InternalPromptInputProps extends PromptInputProps, InternalBaseComponentProps {}
+
+const InternalPromptInput = React.forwardRef(
+  (
+    {
+      value,
+      actionButtonAriaLabel,
+      actionButtonIconName,
+      actionButtonIconUrl,
+      actionButtonIconSvg,
+      actionButtonIconAlt,
+      ariaLabel,
+      autoComplete = true,
+      autoFocus,
+      disableActionButton,
+      disableBrowserAutocorrect,
+      disabled,
+      maxRows,
+      minRows,
+      name,
+      onAction,
+      onBlur,
+      onChange,
+      onFocus,
+      onKeyDown,
+      onKeyUp,
+      placeholder,
+      readOnly,
+      spellcheck,
+      __internalRootRef = null,
+      ...rest
+    }: InternalPromptInputProps,
+    ref: Ref<PromptInputProps.Ref>
+  ) => {
+    const { ariaLabelledby, ariaDescribedby, controlId, invalid, warning } = useFormFieldContext(rest);
+
+    const baseProps = getBaseProps(rest);
+
+    const textareaRef = useRef<HTMLTextAreaElement>(null);
+
+    const isCompactMode = useDensityMode(textareaRef) === 'compact';
+
+    useForwardFocus(ref, textareaRef);
+
+    const handleKeyDown = (event: React.KeyboardEvent<HTMLTextAreaElement>) => {
+      if (onKeyDown) {
+        fireKeyboardEvent(onKeyDown, event);
+      }
+
+      if (event.key === 'Enter' && !event.shiftKey) {
+        if ('form' in event.target && event.target.form !== null && !event.isDefaultPrevented()) {
+          (event.target.form as HTMLFormElement).requestSubmit();
+        }
+
+        event.preventDefault();
+        fireNonCancelableEvent(onAction, { value });
+      }
+    };
+
+    const handleChange = (event: React.ChangeEvent<HTMLTextAreaElement>) => {
+      fireNonCancelableEvent(onChange, { value: event.target.value });
+      adjustTextareaHeight();
+    };
+
+    const adjustTextareaHeight = useCallback(() => {
+      const PADDING = 4;
+      const LINE_HEIGHT = 20;
+      if (textareaRef.current) {
+        textareaRef.current.style.height = 'auto';
+
+        const newHeight = Math.min(
+          textareaRef.current.scrollHeight + PADDING,
+          (maxRows ?? 3) * (LINE_HEIGHT + PADDING / 2) + PADDING
```

Comment:
suggest putting this default value of `3` where we destructure the props rather than inline like this, it makes it easier to see as an overview